lib/catalog.rb

```diff
@@ -1,43 +1,16 @@
 require 'lib/product'
 
 class Catalog
-  @db = Hash[[{
-    id: 1,
-    name: 'Agile Web Development with Rails 5',
-    price: 2800,
-    vat_category_id: 2
-  }, {
-    id: 2,
-    name: 'Data Science Essentials in Python',
-    price: 1900,
-    vat_category_id: 2
-  }, {
-    id: 3,
-    name: 'Web Development with Clojure, Second Edition',
-    price: 2400,
-    vat_category_id: 2
-  }, {
-    id: 4,
-    name: 'Serverless Single Page Apps',
-    price: 3000,
-    vat_category_id: 2
-  }, {
-    id: 5,
-    name: 'Deploying with JRuby 9k',
-    price: 1600,
-    vat_category_id: 2
-  }, {
-    id: 6,
-    name: 'Pragmatic T-Shirt',
-    price: 900,
-    vat_category_id: 1
-  }].map { |attrs| [attrs[:id], Product.new(attrs)] }].freeze
+  def initialize(products = [])
+    raise(ArgumentError) if products.nil?
+    @products = Hash[products.map { |product| [product.id, product] }].freeze
```

Comment:
Maybe this will help you:

```ruby
>> products = [{id: 1, name: 'foo'}, {id: 2, name: 'bar'}]
=> [{:id=>1, :name=>"foo"}, {:id=>2, :name=>"bar"}]
>> products.inject({}){|hash, product| hash[product[:id]] = product; hash }
=> {1=>{:id=>1, :name=>"foo"}, 2=>{:id=>2, :name=>"bar"}}
```

you do not reduce anything just change then `inject` will fit better. In inject (reduce) you always need return self (hash in my example) for next iternation